packages/@aws-cdk/pipelines/lib/stage.ts

```diff
@@ -224,6 +302,53 @@ export class CdkStage extends CoreConstruct {
     return stripPrefix(s, `${this.stageName}-`);
   }
 
+  /**
+   * Add a security check before the prepare/deploy actions of an CDK stage.
+   * The security check consists of two actions:
+   *  - CodeBuild Action to check for security changes in a stage
+   *  - Manual Approval Action that is auto approved via a Lambda if no security changes detected
+   */
+  private addSecurityCheck(appStageName: string, options?: BaseStageOptions) {
+    const { cdkDiffProject } = this.getApplicationSecurityCheck();
+
+    const approveActionName = `${appStageName}ManualApproval`;
+    const diffAction = new CodeBuildAction({
+      runOrder: this.nextSequentialRunOrder(),
+      actionName: `${appStageName}SecurityCheck`,
+      input: this.cloudAssemblyArtifact,
+      project: cdkDiffProject,
+      variablesNamespace: `${appStageName}SecurityCheck`,
+      environmentVariables: {
+        STACK_NAME: {
+          value: this.pipelineStage.pipeline.stack.stackName,
+          type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
+        },
+        STAGE_NAME: {
+          value: this.stageName,
+          type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
+        },
+        ACTION_NAME: {
+          value: approveActionName,
+          type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
+        },
+      },
+    });
+
+    const approve = new cpactions.ManualApprovalAction({
+      actionName: approveActionName,
+      runOrder: this.nextSequentialRunOrder(),
+      additionalInformation: `#{${appStageName}SecurityCheck.MESSAGE}`,
+      // Boot strap cicd2 us-west-1 --trust
+      externalEntityLink: `https://#{${appStageName}SecurityCheck.LINK}`,
+      notificationTopic: (options?.securityNotifyEmails !== []
+        ? options?.securityNotificationTopic ?? this.securityNotificationTopic
+        : undefined),
+      notifyEmails: options?.securityNotifyEmails ?? this.securityNotifyEmails,
```

Comment:
I am fully aware this piece of code is suspicious AF. However, I don't know how we can introduce SNS Topics any other way.

**Constraints**
`pipeline.addApplicationStage()` uses the same options as `stage.addApplication()`. **HOWEVER** the semantics behind the usage of options are implicitly different.

Configuring an option in `pipeline.addApplicationStage()` inherently implies that the entire stage will behave  according to that option. Where as configuring an option in `stage.addApplication()` implies that only the application itself will be configure the option.

Interestingly enough, the current options `AddStageOptions.manualApprovals` and `AddStageOptions.extraRunOrderSpace` **only** affect the first application if configured with `pipeline.addApplicationStage()`. IMO this behavior doesn't make sense semantically. 

I see two ways out of this:
* Explicitly state in the doc strings in `pipeline.addApplicationStage()` that those options will only apply to the application added.
* Support Stage wide 'default' configurations that can be overridden in function calls. Following a notion of 'specificity'. Stage wide configurations are less specific than action configurations

Reply:
@skinny85 @rix0rrr would love some thoughts on this when you two have the time